File: src/07.greedy/01.arrayGreedy/18.169.ts
export function majorityElement(nums: number[]): number {
    /**
    //  笨办法
    const n = nums.length;
    const m = {};
    for (let i = 0; i < n; i++) {
        if (!m[nums[i]]) {
            m[nums[i]] = 1;
        } else {
            m[nums[i]]++;
        }
    }

    let max = -Infinity;
    let maxK = 0;
    for (let k in m) {
        if (m[k] > n / 2) {
            if (m[k] > max) {
                max = m[k];
                maxK = Number(k);
            }
        }
    }

    return maxK;
     
     */

    // 摩尔投票法, 满足 时间复杂度 O(n), 空间复杂度 O(1)
    let candNum = nums[0];
    let count = 1;
    for (let i = 1; i < nums.length; i++) {
        // 票数为 0
        if (count === 0) {
            // 切换元素
            candNum = nums[i];
        }

        // 如果元素相同, 票数 +1, 反之票数 -1
        if (candNum === nums[i]) {
            count += 1;
        } else {
            count -= 1;
        }
    }

    return candNum;
}

File: src/07.greedy/01.arrayGreedy/18.169.go
package array_greedy

func MajorityElement(nums []int) int {
	// 多数投票法

	// 候选人
	major := 0

	// 票数
	count := 0

	for _, v := range nums {

		// 如果票数为 0, 切换候选人
		if count == 0 {
			major = v
		}

		// 候选人相同, 票数 +1, 反之 -1
		if major == v {
			count += 1
		} else {
			count -= 1
		}
	}

	return major
}
